src/librustc_mir/borrow_check/nll/region_infer/mod.rs

```diff
@@ -465,37 +465,53 @@ impl<'tcx> RegionInferenceContext<'tcx> {
         // constraints we have accumulated.
         let mut inferred_values = self.liveness_constraints.clone();
 
-        while changed {
-            changed = false;
-            debug!("propagate_constraints: --------------------");
-            for constraint in &self.constraints {
-                debug!("propagate_constraints: constraint={:?}", constraint);
-
-                // Grow the value as needed to accommodate the
-                // outlives constraint.
-                let Ok(made_changes) = self.dfs(
-                    mir,
-                    CopyFromSourceToTarget {
-                        source_region: constraint.sub,
-                        target_region: constraint.sup,
-                        inferred_values: &mut inferred_values,
-                        constraint_point: constraint.point,
-                        constraint_span: constraint.span,
-                    },
-                );
+        let dependency_map = self.build_dependency_map();
+        let mut dirty_list = self.constraints.clone();
+
+        debug!("propagate_constraints: --------------------");
+        while let Some(constraint) = dirty_list.pop() {
+            debug!("propagate_constraints: constraint={:?}", constraint);
 
-                if made_changes {
-                    debug!("propagate_constraints:   sub={:?}", constraint.sub);
-                    debug!("propagate_constraints:   sup={:?}", constraint.sup);
-                    changed = true;
+            // Grow the value as needed to accommodate the
+            // outlives constraint.
+            let Ok(made_changes) = self.dfs(
+                mir,
+                CopyFromSourceToTarget {
+                    source_region: constraint.sub,
+                    target_region: constraint.sup,
+                    inferred_values: &mut inferred_values,
+                    constraint_point: constraint.point,
+                    constraint_span: constraint.span,
+                },
+            );
+
+            if made_changes {
+                debug!("propagate_constraints:   sub={:?}", constraint.sub);
+                debug!("propagate_constraints:   sup={:?}", constraint.sup);
+
+                if let Some(constraint_deps) = dependency_map.get(&constraint.sup) {
+                    for dep_idx in constraint_deps {
```

Comment:
Now this can be

```rust
if let Some(constraint_deps) = dependency_map.get(&constraint.sup) {
    dirty_list.extend(constraint_deps);
}
```

Not that it matters much, but generally invoking `extend` is better than `for` loop that calls `push`. This is because sometimes the iterator knows how long it will be, which lets `dirty_list` reallocate its backing array more efficiently.

Finally, you *could* simplify this further, like so:

```rust
dirty_list.extend(
    dependency_map.get(&constraint.sup).unwrap_or(&vec![])
);
```

but I'm not sure that's really *clearer* (arguably less so). The key here is that creating an empty vector (i.e., `vec![]`) is basically free -- it's just some null values, no allocation or anything. In the case where I have a "map to vec", I often do this style, under the idea that a missing entry is equivalent to an empty vector. But either way really.